File: pset2/crack/crack.c
// Decrypt a hashed password that use C's DES encryption.
#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <cs50.h>

string hash;
char salt[3];
int hashMatches(string word);

int main(int argc, string argv[])
{
    // Ensure user input exist, otherwise return an error and exit.
    if (argc != 2)
    {
        printf("Usage: ./crack <password>\n");
        return 1;
    }

char word[5];
hash = argv[1];
salt[0] = hash[0];
salt[1] = hash[1];

// Store all ASCII characters in an array
string letters = "\0abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

// Use a reverse loop to brute force all possible hashes.

for (int m = 0; m < 57; m++)
    for (int l = 0; l < 57; l++)
        for (int k = 0; k < 57; k++)
            for (int j = 0; j < 57; j++)
                for (int i = 0; i < 57; i++)
                {
                    word[0] = letters[i];
                    word[1] = letters[j];
                    word[2] = letters[k];
                    word[3] = letters[l];
                    word[4] = letters[m];
                    if (hashMatches(word)) return 0;
                }
    return 1;
}

// Compare the generated hash against user's input of hash.
int hashMatches(string word)
{
    if (strcmp(crypt(word, salt), hash) == 0)
    {
        printf("%s\n", word);
        return 1;
    }
    return 0;
}

File: pset6/crack/crack.py
# Port of crack.c to Python
import crypt
import string
from sys import argv

# Ensure the user entered at least one arg, otherwise exit
if (len(argv) != 2):
    print("Usage: python crack.py hash")
    exit(1)

# init a hash as as a string
hashed = str(argv[1])
salt = str(hashed[:2])
password = ""

# Store all ASCII charactersin an tuple
letters = string.ascii_letters
letlen = len(letters)

# iterate through a hash to brute force all possible hashes
for i in range(letlen):
    for j in range(letlen):
        for k in range(letlen):
            for m in range(letlen):
                for n in range(letlen):
                    # check if the hash matches the word and exits if passed
                    if crypt.crypt(password, salt) == hashed:
                        print(password)
                        exit(0)
                    password = letters[n] + password[1:]
                password = password[:1] + letters[m] + password[2:]
            password = password[:2] + letters[k] + password[3:]
        password = password[:3] + letters[j] + password[4:]
    password = password[:4] + letters[i] + password[5:]